test/performance/scale_from_zero_test.go

```diff
@@ -0,0 +1,198 @@
+	domain         string
+}
+
+func setupClients(t *testing.T) *test.Clients {
+	clients, err := test.NewClients(
+		pkgTest.Flags.Kubeconfig,
+		pkgTest.Flags.Cluster,
+		test.ServingNamespace)
+	if err != nil {
+		t.Fatalf("Couldn't initialize clients: %v", err)
+	}
+	return clients
+}
+
+func tearDown(clients *test.Clients, names test.ResourceNames, logger *logging.BaseLogger) {
+	if clients != nil && clients.ServingClient != nil {
+		clients.ServingClient.Delete([]string{names.Route}, []string{names.Config}, []string{names.Service})
+	}
+}
+
+func createRouteAndConfig(clients *test.Clients, logger *logging.BaseLogger, imagePath string, options *test.Options) (test.ResourceNames, error) {
+	var names test.ResourceNames
+	names.Config = test.AppendRandomString(configName, logger)
+	names.Route = test.AppendRandomString(routeName, logger)
+
+	if err := test.CreateConfiguration(logger, clients, names, imagePath, options); err != nil {
+		return test.ResourceNames{}, err
+	}
+	err := test.CreateRoute(logger, clients, names)
+	return names, err
+}
+
+func isDeploymentScaledToZero() func(d *v1beta1.Deployment) (bool, error) {
+	return func(d *v1beta1.Deployment) (bool, error) {
+		return d.Status.ReadyReplicas == 0, nil
+	}
+}
+
+func setup(t *testing.T, clients *test.Clients, name string) *testContext {
+	logger := logging.GetContextLogger("TestScaleFromZero (" + name + ")")
+	imagePath := test.ImagePath("helloworld")
+	names, err := createRouteAndConfig(clients, logger, imagePath, &test.Options{})
+	if err != nil {
+		t.Fatalf("Failed to create Route and Configuration: %v", err)
+	}
+	test.CleanupOnInterrupt(func() { tearDown(clients, names, logger) }, logger)
+
+	logger.Info("Waiting for route to be marked ready.")
+	if err := test.WaitForRouteState(clients.ServingClient, names.Route, test.IsRouteReady, "RouteIsReady"); err != nil {
+		t.Fatalf("The Route %s was not marked as Ready to serve traffic: %v", names.Route, err)
+	}
+
+	config, err := clients.ServingClient.Configs.Get(names.Config, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Configuration %s was not updated with the new revision: %v", names.Config, err)
+	}
+	deploymentName := config.Status.LatestCreatedRevisionName + "-deployment"
+
+	route, err := clients.ServingClient.Routes.Get(names.Route, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Error fetching Route %s: %v", names.Route, err)
+	}
+	domain := route.Status.Domain
+
+	return &testContext{
+		deploymentName: deploymentName,
+		logger:         logger,
+		names:          names,
+		domain:         domain,
+	}
+}
+
+func runScaleFromZero(t *testing.T, clients *test.Clients, name string) time.Duration {
+	ctx := setup(t, clients, name)
+	defer tearDown(clients, ctx.names, ctx.logger)
+
+	ctx.logger.Info("Waiting for deployment to scale to zero.")
+	if err := test.WaitForDeploymentState(
+		clients.KubeClient,
+		ctx.deploymentName,
+		isDeploymentScaledToZero(),
+		"DeploymentScaledToZero",
+		2*time.Minute); err != nil {
```

Comment:
The 2 minutes are probably derived from the default `scale-to-zero-grace-period`, should we update the k-native deployment before the test by setting this param to the min value to reduce the test turnaround time?

Reply:
Good find, I should definitely lower this.

Reply:
I played with this - this needs to be >1minute to avoid timeouts so I think 2min is probably a good spot.